core/code/entity_decode.js

```diff
@@ -89,53 +102,64 @@ function summaryPortalData(a) {
   };
 };
 
-var DETAILED_PORTAL_DATA_LENGTH = SUMMARY_PORTAL_DATA_LENGTH+4;
-
-
-window.decodeArray.portalSummary = function(a) {
-  if (!a) return undefined;
-
-  if (a[0] !== 'p') {
-    throw new Error('Error: decodeArray.portalSUmmary - not a portal');
+function detailsPortalData(a) {
+  return {
+    mods:           a[SUMMARY_PORTAL_DATA_LENGTH+0].map(parseMod),
+    resonators:     a[SUMMARY_PORTAL_DATA_LENGTH+1].map(parseResonator),
+    owner:          a[SUMMARY_PORTAL_DATA_LENGTH+2],
+    artifactDetail: parseArtifactDetail(a[SUMMARY_PORTAL_DATA_LENGTH+3])
   }
+};
 
-  if (a.length == CORE_PORTA_DATA_LENGTH) {
-    return corePortalData(a);
+function extendedPortalData(a) {
+  return {
+    history: parseHistoryDetail(a[DETAILED_PORTAL_DATA_LENGTH] || 0),
   }
+};
 
-  // NOTE: allow for either summary or detailed portal data to be passed in here, as details are sometimes
-  // passed into code only expecting summaries
-  if (a.length != SUMMARY_PORTAL_DATA_LENGTH && a.length != DETAILED_PORTAL_DATA_LENGTH) {
-    log.warn('Portal summary length changed - portal details likely broken!');
-    debugger;
-  }
 
-  return $.extend(corePortalData(a), summaryPortalData(a));
-}
+var dataLen = {
+  core: [CORE_PORTAL_DATA_LENGTH],
+  summary: [SUMMARY_PORTAL_DATA_LENGTH],
+  detailed: [DETAILED_PORTAL_DATA_LENGTH],
+  extended: [EXTENDED_PORTAL_DATA_LENGTH, SUMMARY_PORTAL_DATA_LENGTH]
+};
 
-window.decodeArray.portalDetail = function(a) {
+window.decodeArray.portal = function(a, details) {
   if (!a) return undefined;
 
   if (a[0] !== 'p') {
-    throw new Error('Error: decodeArray.portalDetail - not a portal');
+    throw new Error('Error: decodeArray.portalSUmmary - not a portal');
   }
 
-  if (a.length != DETAILED_PORTAL_DATA_LENGTH) {
-    log.warn('Portal detail length changed - portal details may be wrong');
-    debugger;
+  if (details) {
+    var expected = dataLen[details];
+    if (expected.indexOf(a.length) === -1) {
+      log.warn('Unexpected portal data length: ' + a.length + ' ('+details+')');
+      debugger;
+    }
   }
 
-  //TODO look at the array values, make a better guess as to which index the mods start at, rather than using the hard-coded SUMMARY_PORTAL_DATA_LENGTH constant
+  var data = corePortalData(a);
 
+  if (a.length >= SUMMARY_PORTAL_DATA_LENGTH) {
+    $.extend(data, summaryPortalData(a));
+  }
+  if (a.length >= DETAILED_PORTAL_DATA_LENGTH) {
+    if (details !== 'extended' || a[SUMMARY_PORTAL_DATA_LENGTH]) {
```

Comment:
why not parse it whenever there is data to parse (not `null`)?
```suggestion
    if (a[SUMMARY_PORTAL_DATA_LENGTH]) {
```

Reply:
It is opposite: parsing is forced even if there is no data.
Reason: throw error to make us know about weird data case.

Reply:
Alternatively, we could use `console.warn` here.
@le-jeu?

Reply:
hm, the code is likely to raise an exception if `a[SUMMARY_PORTAL_DATA_LENGTH]` is null due to `parseDetails` assumming it is an Array

Reply:
Exactly.
So question: is it acceptable (if intel changes data - we adapt).
Or better avoid exception, but send just warning.

Exception is better because it's harder to miss.
But warning is better because app is not breaking for everyone.

Reply:
better warn and don't parse than parse and crash in the middle of something
https://github.com/IITC-CE/ingress-intel-total-conversion/pull/436/files/d4834db670a9177471a7834ee18785c48558705a#r574442530